File: core/hanko/abi.ts
/**
 * Direct ABI codec for the Hanko envelope tuple
 * `tuple(bytes32[],bytes,tuple(bytes32,uint256[],uint256[],uint256,uint32,uint32,uint32)[])`.
 *
 * Byte-identical to `AbiCoder.encode/decode` of that type (differential test in
 * `__tests__/security/hanko/hanko-abi.test.ts`). The generic coder parses
 * fragments, wraps every value in Typed/Result objects and copies bytes per
 * word; a busy hub encodes one envelope per certified output, which made the
 * generic path ~5% of hub CPU. Values are already validated by the caller;
 * this module only lays out words. Decoding is bounds-checked and the caller
 * still enforces canonical form by re-encoding.
 */

const WORD = 64; // hex chars per 32-byte word
const HEX_WORD_RE = /^[0-9a-f]{64}$/;

export type HankoAbiClaim = readonly [
  entityId: string,
  entityIndexes: readonly bigint[],
  weights: readonly bigint[],
  threshold: bigint,
  boardChangeDelay: bigint,
  controlChangeDelay: bigint,
  dividendChangeDelay: bigint,
];

export type HankoAbiEnvelope = readonly [
  placeholders: readonly string[],
  packedSignatures: string,
  claims: readonly HankoAbiClaim[],
];

const uintWord = (value: bigint): string => value.toString(16).padStart(WORD, '0');
const bytes32Word = (value: string): string => value.slice(2).toLowerCase();

const uintArray = (values: readonly bigint[]): string => {
  let out = uintWord(BigInt(values.length));
  for (let index = 0; index < values.length; index += 1) out += uintWord(values[index]!);
  return out;
};

const encodeClaim = (claim: HankoAbiClaim): string => {
  const [entityId, indexes, weights, threshold, board, control, dividend] = claim;
  const indexesOffset = 7 * 32;
  const weightsOffset = indexesOffset + 32 * (1 + indexes.length);
  return bytes32Word(entityId)
    + uintWord(BigInt(indexesOffset))
    + uintWord(BigInt(weightsOffset))
    + uintWord(threshold)
    + uintWord(board)
    + uintWord(control)
    + uintWord(dividend)
    + uintArray(indexes)
    + uintArray(weights);
};

/** `0x`-prefixed lowercase hex, identical to `AbiCoder.encode([HANKO_TYPE], [envelope])`. */
export const encodeHankoAbi = (envelope: HankoAbiEnvelope): string => {
  const [placeholders, packed, claims] = envelope;
  const packedHex = packed.slice(2).toLowerCase();
  const packedBytes = packedHex.length / 2;
  const packedPadded = packedHex + '0'.repeat((WORD - (packedHex.length % WORD)) % WORD);

  let placeholdersEnc = uintWord(BigInt(placeholders.length));
  for (let index = 0; index < placeholders.length; index += 1) {
    placeholdersEnc += bytes32Word(placeholders[index]!);
  }
  const packedEnc = uintWord(BigInt(packedBytes)) + packedPadded;

  const claimEncodings = claims.map(encodeClaim);
  let claimsHeads = '';
  let claimsTails = '';
  let claimOffset = 32 * claims.length;
  for (let index = 0; index < claimEncodings.length; index += 1) {
    const encoded = claimEncodings[index]!;
    claimsHeads += uintWord(BigInt(claimOffset));
    claimsTails += encoded;
    claimOffset += encoded.length / 2;
  }
  const claimsEnc = uintWord(BigInt(claims.length)) + claimsHeads + claimsTails;

  const placeholdersOffset = 3 * 32;
  const packedOffset = placeholdersOffset + placeholdersEnc.length / 2;
  const claimsOffset = packedOffset + packedEnc.length / 2;
  return '0x'
    + uintWord(32n)
    + uintWord(BigInt(placeholdersOffset))
    + uintWord(BigInt(packedOffset))
    + uintWord(BigInt(claimsOffset))
    + placeholdersEnc
    + packedEnc
    + claimsEnc;
};

class HankoAbiReader {
  constructor(private readonly hex: string) {}

  /** Read the 32-byte word at byte offset `offset`; throws when out of range. */
  word(offset: number): string {
    const start = offset * 2;
    if (!Number.isSafeInteger(offset) || offset < 0 || start + WORD > this.hex.length) {
      throw new Error(`HANKO_ABI_OUT_OF_BOUNDS:${offset}`);
    }
    return this.hex.slice(start, start + WORD);
  }

  uint(offset: number): bigint {
    return BigInt(`0x${this.word(offset)}`);
  }

  /** Small non-negative integer (lengths / offsets) — anything past the buffer is invalid. */
  size(offset: number): number {
    const value = this.uint(offset);
    if (value > BigInt(this.hex.length / 2)) throw new Error(`HANKO_ABI_SIZE_INVALID:${offset}`);
    return Number(value);
  }

  bytes32(offset: number): string {
    const word = this.word(offset);
    if (!HEX_WORD_RE.test(word)) throw new Error(`HANKO_ABI_WORD_INVALID:${offset}`);
    return `0x${word}`;
  }

  uintArray(offset: number): bigint[] {
    const length = this.size(offset);
    const values: bigint[] = new Array(length);
    for (let index = 0; index < length; index += 1) values[index] = this.uint(offset + 32 * (1 + index));
    return values;
  }

  /** AbiCoder masks fixed-width integers to their size; the canonical re-encode rejects the excess. */
  uint32(offset: number): bigint {
    return this.uint(offset) & 0xffff_ffffn;
  }

  bytes(offset: number): string {
    const length = this.size(offset);
    const start = (offset + 32) * 2;
    const end = start + length * 2;
    if (end > this.hex.length) throw new Error(`HANKO_ABI_OUT_OF_BOUNDS:${offset}`);
    return `0x${this.hex.slice(start, end)}`;
  }
}

/**
 * Decode `AbiCoder.encode([HANKO_TYPE], [envelope])`. Input must be `0x` +
 * even-length lowercase hex (the caller normalises). Structural violations
 * throw `HANKO_ABI_*`; like AbiCoder, trailing bytes and oversized fixed-width
 * integers decode leniently — the caller re-encodes to reject every
 * non-canonical layout.
 */
export const decodeHankoAbi = (encoded: string): HankoAbiEnvelope => {
  const hex = encoded.slice(2);
  const reader = new HankoAbiReader(hex);
  const tuple = reader.size(0);
  const placeholdersAt = tuple + reader.size(tuple);
  const packedAt = tuple + reader.size(tuple + 32);
  const claimsAt = tuple + reader.size(tuple + 64);

  const placeholderCount = reader.size(placeholdersAt);
  const placeholders: string[] = new Array(placeholderCount);
  for (let index = 0; index < placeholderCount; index += 1) {
    placeholders[index] = reader.bytes32(placeholdersAt + 32 * (1 + index));
  }
  const packed = reader.bytes(packedAt);

  const claimCount = reader.size(claimsAt);
  const claimsBase = claimsAt + 32;
  const claims: HankoAbiClaim[] = new Array(claimCount);
  for (let index = 0; index < claimCount; index += 1) {
    const claimAt = claimsBase + reader.size(claimsBase + 32 * index);
    claims[index] = [
      reader.bytes32(claimAt),
      reader.uintArray(claimAt + reader.size(claimAt + 32)),
      reader.uintArray(claimAt + reader.size(claimAt + 64)),
      reader.uint(claimAt + 96),
      reader.uint32(claimAt + 128),
      reader.uint32(claimAt + 160),
      reader.uint32(claimAt + 192),
    ];
  }
  return [placeholders, packed, claims];
};

File: rscore/crates/hanko/src/abi.rs
//! Direct ABI codec for the Hanko envelope tuple
//! `tuple(bytes32[],bytes,tuple(bytes32,uint256[],uint256[],uint256,uint32,uint32,uint32)[],bytes[])`
//! (HankoVerifier.HankoBytes: placeholders, packedSignatures, claims,
//! memberSignatures).
//!
//! Parity target: core/hanko/abi.ts. Byte-identical to what `AbiCoder` writes
//! for that type. Decoding is bounds-checked and, like TypeScript, lenient
//! about trailing bytes and oversized fixed-width integers — the caller
//! rejects every non-canonical layout by re-encoding.

use crate::HankoError;

pub type Word = [u8; 32];

const WORD: usize = 32;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct AbiClaim {
    pub entity_id: Word,
    pub entity_indexes: Vec<Word>,
    pub weights: Vec<Word>,
    pub threshold: Word,
    pub board_change_delay: u32,
    pub control_change_delay: u32,
    pub dividend_change_delay: u32,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct AbiEnvelope {
    pub placeholders: Vec<Word>,
    pub packed_signatures: Vec<u8>,
    pub claims: Vec<AbiClaim>,
    /// Aligned with `placeholders` (or empty): an ERC-1271 proof for the
    /// placeholder at the same index. Only the jurisdiction can evaluate one.
    pub member_signatures: Vec<Vec<u8>>,
}

pub fn word_from_usize(value: usize) -> Word {
    let mut word = [0_u8; 32];
    word[24..].copy_from_slice(&(value as u64).to_be_bytes());
    word
}

fn word_from_u32(value: u32) -> Word {
    let mut word = [0_u8; 32];
    word[28..].copy_from_slice(&value.to_be_bytes());
    word
}

fn encode_claim(claim: &AbiClaim, output: &mut Vec<u8>) {
    let indexes_offset = 7 * WORD;
    let weights_offset = indexes_offset + WORD * (1 + claim.entity_indexes.len());
    output.extend_from_slice(&claim.entity_id);
    output.extend_from_slice(&word_from_usize(indexes_offset));
    output.extend_from_slice(&word_from_usize(weights_offset));
    output.extend_from_slice(&claim.threshold);
    output.extend_from_slice(&word_from_u32(claim.board_change_delay));
    output.extend_from_slice(&word_from_u32(claim.control_change_delay));
    output.extend_from_slice(&word_from_u32(claim.dividend_change_delay));
    for values in [&claim.entity_indexes, &claim.weights] {
        output.extend_from_slice(&word_from_usize(values.len()));
        for value in values {
            output.extend_from_slice(value);
        }
    }
}

/// ABI `bytes`: length word + payload right-padded to a word boundary.
fn encode_bytes(value: &[u8], output: &mut Vec<u8>) {
    let padding = (WORD - value.len() % WORD) % WORD;
    output.extend_from_slice(&word_from_usize(value.len()));
    output.extend_from_slice(value);
    output.resize(output.len() + padding, 0);
}

/// ABI `bytes[]`: length word, one offset word per item, then each `bytes`.
fn encode_bytes_array(values: &[Vec<u8>]) -> Vec<u8> {
    let mut heads = Vec::with_capacity(WORD * values.len());
    let mut tails = Vec::new();
    let mut offset = WORD * values.len();
    for value in values {
        heads.extend_from_slice(&word_from_usize(offset));
        let before = tails.len();
        encode_bytes(value, &mut tails);
        offset += tails.len() - before;
    }
    let mut output = Vec::with_capacity(WORD + heads.len() + tails.len());
    output.extend_from_slice(&word_from_usize(values.len()));
    output.extend_from_slice(&heads);
    output.extend_from_slice(&tails);
    output
}

pub fn encode_hanko_abi(envelope: &AbiEnvelope) -> Vec<u8> {
    let mut placeholders = Vec::with_capacity(WORD * (1 + envelope.placeholders.len()));
    placeholders.extend_from_slice(&word_from_usize(envelope.placeholders.len()));
    for placeholder in &envelope.placeholders {
        placeholders.extend_from_slice(placeholder);
    }

    let mut packed = Vec::with_capacity(2 * WORD + envelope.packed_signatures.len());
    encode_bytes(&envelope.packed_signatures, &mut packed);

    let mut heads = Vec::with_capacity(WORD * envelope.claims.len());
    let mut tails = Vec::new();
    let mut claim_offset = WORD * envelope.claims.len();
    for claim in &envelope.claims {
        heads.extend_from_slice(&word_from_usize(claim_offset));
        let before = tails.len();
        encode_claim(claim, &mut tails);
        claim_offset += tails.len() - before;
    }
    let mut claims = Vec::with_capacity(WORD + heads.len() + tails.len());
    claims.extend_from_slice(&word_from_usize(envelope.claims.len()));
    claims.extend_from_slice(&heads);
    claims.extend_from_slice(&tails);

    let members = encode_bytes_array(&envelope.member_signatures);

    let placeholders_offset = 4 * WORD;
    let packed_offset = placeholders_offset + placeholders.len();
    let claims_offset = packed_offset + packed.len();
    let members_offset = claims_offset + claims.len();
    let mut output = Vec::with_capacity(
        5 * WORD + placeholders.len() + packed.len() + claims.len() + members.len(),
    );
    output.extend_from_slice(&word_from_usize(WORD));
    output.extend_from_slice(&word_from_usize(placeholders_offset));
    output.extend_from_slice(&word_from_usize(packed_offset));
    output.extend_from_slice(&word_from_usize(claims_offset));
    output.extend_from_slice(&word_from_usize(members_offset));
    output.extend_from_slice(&placeholders);
    output.extend_from_slice(&packed);
    output.extend_from_slice(&claims);
    output.extend_from_slice(&members);
    output
}

struct AbiReader<'a> {
    bytes: &'a [u8],
}

impl<'a> AbiReader<'a> {
    fn word(&self, offset: usize) -> Result<Word, HankoError> {
        let end = offset
            .checked_add(WORD)
            .ok_or(HankoError::AbiOutOfBounds(offset))?;
        if end > self.bytes.len() {
            return Err(HankoError::AbiOutOfBounds(offset));
        }
        let mut word = [0_u8; 32];
        word.copy_from_slice(&self.bytes[offset..end]);
        Ok(word)
    }

    /// A length or offset. Anything past the buffer is invalid, which also
    /// keeps every later addition inside `usize`.
    fn size(&self, offset: usize) -> Result<usize, HankoError> {
        let word = self.word(offset)?;
        if word[..24].iter().any(|byte| *byte != 0) {
            return Err(HankoError::AbiSizeInvalid(offset));
        }
        let value = u64::from_be_bytes(word[24..].try_into().expect("8 bytes")) as usize;
        if value > self.bytes.len() {
            return Err(HankoError::AbiSizeInvalid(offset));
        }
        Ok(value)
    }

    fn word_array(&self, offset: usize) -> Result<Vec<Word>, HankoError> {
        let length = self.size(offset)?;
        let mut values = Vec::with_capacity(length);
        for index in 0..length {
            values.push(self.word(offset + WORD * (1 + index))?);
        }
        Ok(values)
    }

    /// AbiCoder masks fixed-width integers to their size; the canonical
    /// re-encode rejects the excess.
    fn uint32(&self, offset: usize) -> Result<u32, HankoError> {
        let word = self.word(offset)?;
        Ok(u32::from_be_bytes(word[28..].try_into().expect("4 bytes")))
    }

    fn bytes(&self, offset: usize) -> Result<Vec<u8>, HankoError> {
        let length = self.size(offset)?;
        let start = offset + WORD;
        let end = start
            .checked_add(length)
            .ok_or(HankoError::AbiOutOfBounds(offset))?;
        if end > self.bytes.len() {
            return Err(HankoError::AbiOutOfBounds(offset));
        }
        Ok(self.bytes[start..end].to_vec())
    }
}

pub fn decode_hanko_abi(encoded: &[u8]) -> Result<AbiEnvelope, HankoError> {
    let reader = AbiReader { bytes: encoded };
    let tuple = reader.size(0)?;
    let placeholders_at = tuple + reader.size(tuple)?;
    let packed_at = tuple + reader.size(tuple + WORD)?;
    let claims_at = tuple + reader.size(tuple + 2 * WORD)?;
    let members_at = tuple + reader.size(tuple + 3 * WORD)?;

    let placeholders = reader.word_array(placeholders_at)?;
    let packed_signatures = reader.bytes(packed_at)?;

    let claim_count = reader.size(claims_at)?;
    let claims_base = claims_at + WORD;
    let mut claims = Vec::with_capacity(claim_count);
    for index in 0..claim_count {
        let claim_at = claims_base + reader.size(claims_base + WORD * index)?;
        let indexes_at = claim_at + reader.size(claim_at + WORD)?;
        let weights_at = claim_at + reader.size(claim_at + 2 * WORD)?;
        claims.push(AbiClaim {
            entity_id: reader.word(claim_at)?,
            entity_indexes: reader.word_array(indexes_at)?,
            weights: reader.word_array(weights_at)?,
            threshold: reader.word(claim_at + 3 * WORD)?,
            board_change_delay: reader.uint32(claim_at + 4 * WORD)?,
            control_change_delay: reader.uint32(claim_at + 5 * WORD)?,
            dividend_change_delay: reader.uint32(claim_at + 6 * WORD)?,
        });
    }

    let member_count = reader.size(members_at)?;
    let members_base = members_at + WORD;
    let mut member_signatures = Vec::with_capacity(member_count);
    for index in 0..member_count {
        let at = members_base + reader.size(members_base + WORD * index)?;
        member_signatures.push(reader.bytes(at)?);
    }
    Ok(AbiEnvelope {
        placeholders,
        packed_signatures,
        claims,
        member_signatures,
    })
}
